homeassistant/components/insteon/utils.py

```diff
@@ -47,91 +64,123 @@
+                ]:
+                    _LOGGER.debug(
+                        "Registering on/off event for %s %d %s",
+                        str(device.address),
+                        group,
+                        event,
+                    )
+                    device.events[group][event].subscribe(
+                        async_fire_group_on_off_event, force_strong_ref=True
+                    )
+
+
+def register_new_device_callback(hass, config):
+    """Register callback for new Insteon device."""
+
     @callback
-    def async_new_insteon_device(device):
+    def async_new_insteon_device(address=None):
         """Detect device from transport to be delegated to platform."""
-        ipdb = IPDB()
-        for state_key in device.states:
-            platform_info = ipdb[device.states[state_key]]
-            if platform_info and platform_info.platform:
-                platform = platform_info.platform
-
-                if platform == "on_off_events":
-                    device.states[state_key].register_updates(_fire_button_on_off_event)
-
-                else:
-                    _LOGGER.info(
-                        "New INSTEON device: %s (%s) %s",
-                        device.address,
-                        device.states[state_key].name,
-                        platform,
-                    )
+        _LOGGER.info(
+            "Adding new INSTEON device to Home Assistant with address %s", address
+        )
+        hass.async_add_job(devices.async_save, hass.config.config_dir)
+        device = devices[address]
+        device.status()
+        platforms = get_device_platforms(device)
+        for platform in platforms:
+            if platform == ON_OFF_EVENTS:
+                add_on_off_event_device(hass, device)
 
-                    hass.async_create_task(
-                        discovery.async_load_platform(
-                            hass,
-                            platform,
-                            DOMAIN,
-                            discovered={
-                                "address": device.address.id,
-                                "state_key": state_key,
-                            },
-                            hass_config=config,
-                        )
+            else:
+                hass.async_create_task(
+                    discovery.async_load_platform(
+                        hass,
+                        platform,
+                        DOMAIN,
+                        discovered={"address": device.address.id},
+                        hass_config=config,
                     )
+                )
 
-    insteon_modem.devices.add_device_callback(async_new_insteon_device)
+    devices.subscribe(async_new_insteon_device, force_strong_ref=True)
 
 
 @callback
-def async_register_services(hass, config, insteon_modem):
+def async_register_services(hass):
     """Register services used by insteon component."""
 
     def add_all_link(service):
```

Comment:
All the services needs to be either async callbacks or coroutine functions if we want async context inside.

Reply:
I have been avoiding async callbacks in services believing that they could not be awaitable or coroutines. Obviously that is wrong. I changed / simplified all of the services to be async that call async methods. Much easier for sure.

So when does a method require the @callback decorator? It seems to only be on non-async methods but what is the purpose? I never know when to use it and when not to. I just use it when I am told by a reviewer.

Reply:
We mark regular functions that must only be run in the event loop with `@callback` and name them `async_*`.

Reply:
When would a regular function need to be run in the event loop? Is that a thread safe concern?